docs/docs/concepts/Sync_Async.md

```diff
@@ -0,0 +1,204 @@
+    var count = 0;
+    var cursor = '0';
+    do{
+        async_client.block((client)=>{
+            var res = client.call('scan', cursor);
+            cursor = res[0];
+            var keys = res[1];
+            keys.forEach((key) => {
+                if (client.call('hget', key, 'name') == expected_name) {
+                    count += 1;
+                }
+            });
+        });
+    } while(cursor != '0');
+    return count;
+});
+```
+
+Both implementations return the same result, but the seconds runs in the background and block Redis just to analyze the next batch of keys that returned from the scan command. Other commands will be processed in between the scan batches. Notice that the Coroutine approach allows the key space to be changed while the scanning it, function writer will need to decide if this is acceptable.
+
+# Start Sync and Move Async
+
+The above example is costly, even though Redis is not blocked it is still takes time to return the reply to the user. If we flatten the requirement in such way that we agree to get an approximate value, we can get a much better performance (on most cases). We will cache the result on a key called `<name>_count` and set some expiration on that key so that we will recalculate the value from time to time. The new code will look like this:
+
+```js
+#!js api_version=1.0 name=lib
+
+redis.registerAsyncFunction('test', async function(async_client, expected_name){
+    // check the cache first
+    var cached_value = async_client.block((client)=>{
+        return client.call('get', expected_name + '_count');
+    });
+
+    if (cached_value != null) {
+        return cached_value;
+    }
+
+
+    var count = 0;
+    var cursor = '0';
+    do{
+        async_client.block((client)=>{
+            var res = client.call('scan', cursor);
+            cursor = res[0];
+            var keys = res[1];
+            keys.forEach((key) => {
+                if (client.call('hget', key, 'name') == expected_name) {
+                    count += 1;
+                }
+            });
+        });
+    } while(cursor != '0');
+
+    // set count to the cache wil 5 seconds expiration
+    async_client.block((client)=>{
+        client.call('set', expected_name + '_count', count);
+        client.call('expire', expected_name + '_count', 5);
+    });
+
+    return count;
+});
+```
+
+The above code works as expected, it first check the cache, if cache exists it returns it, otherwise it is perform the calculation and update the cache. But the above example is not optimal, the callback is a Coroutine which means that it will always be calculated on a background thread. Moving to a background thread by itself is costly, the best approach would have been to check the cache synchronously and only if its not there, move to the background. Triggers and Functions allows to start synchronously and move asynchronously using `executeAsync` function. The new code:
```

Comment:
```suggestion
The above code works as expected. It first checks the cache and if the cache exists it's returned. Otherwise it will perform the calculation and update the cache. But the above example is not optimal. The callback is a Coroutine, which means that it will always be calculated on a background thread. Intrinsically, moving to a background thread is costly. The best approach would be to check the cache synchronously and, only if its not there, move to the background. Triggers and Functions provides for starting synchronously and then moving asynchronously using `executeAsync` function as required. The new code:
```